climada/hazard/tc_tracks.py

```diff
@@ -2867,3 +2868,64 @@ def _zlib_from_dataarray(data_var: xr.DataArray) -> bool:
     if np.issubdtype(data_var.dtype, float) or np.issubdtype(data_var.dtype, int):
         return True
     return False
+
+
+def compute_track_density(
+    tc_track: TCTracks,
+    res: int = 5,
+    time_step: float = 1,
+    density: bool = False,
+    filter_tracks: bool = True,
+) -> tuple[np.ndarray, tuple]:
+    """Compute absolute and normalized tropical cyclone track density. First, the function ensure
+    the same temporal resolution of all tracks by calling :py:meth:`equal_timestep`. Second, it
+    creates 2D bins of the specified resolution (e.g. 1° x 1°). Third, since tracks are not lines
+    but a series of points, it counts the number of points per bin. Lastly, it returns the absolute
+    or normalized count per bin. This function works under the hood of :py:meth:`plot_track_density`
+    but can be used separtly as input data for more sophisticated track density plots.
+    If filter track is True, only a maximum of one point is added to each grid cell for every track.
+    Hence, the resulting density will represent the number of differt tracksper grid cell.
+
+    Parameters:
+    ----------
+    res: int (optional) Default: 5°
+        resolution in degrees of the grid bins in which the density will be computed
+    time_step: float (optional) default: 1h
+        temporal resolution in hours to be apllied to the tracks, to ensure that every track
+        will have the same resolution.
+    density: bool (optional) default: False
+        If False it returns the number of samples in each bin. If True, returns the
+        probability density function at each bin computed as count_bin / tot_count.
+    filter_tracks: bool (optional) default: True
+        If True the track density is computed as the number of different tracks crossing a grid
+        cell. If False, the track density takes into account how long the track stayed in each
+        grid cell. Hence slower tracks increase the density if the parameter is set to False.
+
+    Returns:
+    -------
+    hist: 2D np.ndarray
+        2D matrix containing the track density
+    """
+
+    # ensure equal time step
+    # tc_track.equal_timestep(time_step_h=time_step)
+
+    # Define grid resolution and bounds for density computation
+    lat_bins = np.linspace(-90, 90, int(180 / res))
+    lon_bins = np.linspace(-180, 180, int(360 / res))
+
+    # Compute 2D density
+    hist_count = csr_matrix((len(lat_bins) - 1, len(lon_bins) - 1))
+    for track in tc_track.data:
+
+        # Compute 2D density
+        hist_new, _, _ = np.histogram2d(
+            track.lat.values, track.lon.values, bins=[lat_bins, lon_bins], density=False
+        )
+        hist_new = csr_matrix(hist_new)
+        hist_new[hist_new > 1] = 1 if filter_tracks else hist_new
```

Comment:
why not something like this (seems easier to read):
```
if filter_tracks:
    np.clip(hist_new, a_max=1, out=hist_new)
```

Reply:
You might need to do this before converting to a csr_matrix

Reply:
yes I like it, then I guess it doesn't really make sense to use sparse matrixes anymore. Since the core function  `np.histogram2d` doesn't exist with sparse, and the later operations are better off with numpy.

Reply:
mmh actually np.clip creates some issues later on with the testing, I'll keep it as it is.

Reply:
If you keep it dense, then you should justmake it clear in the docstring that this code does not work well at high resolution. It would be good to do some simple testing (res=1, 0.1, 0.01, 0.001) and see how much memory / time it takes.

Reply:
Or use np.where(x>1,1,x)

Reply:
> If you keep it dense, then you should just make it clear in the docstring that this code does not work well at high resolution. It would be good to do some simple testing (res=1, 0.1, 0.01, 0.001) and see how much memory / time it takes.

At the moment, using sparse (and numpy), with 300 tracks at global scale it takes:

1°: 141ms
0.1°: 9.1s
0.01°: 17 minutes
0.001°: ...

**IF** it makes sense to use this method at really high resolutions (e.g > 0.1°) I will try to optimize it.

Reply:
I do not know whether higher resolution is needed, but it would be good to mention in the docstring what kind of resolution is reasonable.

Reply:
From my experience anything below 0.1 degrees is far too noisy, typically one uses something like 1, 5 or even 10 degrees.